subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/DefaultCacheLockingManager.java

```diff
@@ -39,9 +47,22 @@ public DefaultCacheLockingManager(CacheRepository cacheRepository, ArtifactCache
                 .withCrossVersionCache(CacheBuilder.LockTarget.CacheDirectory)
                 .withDisplayName("artifact cache")
                 .withLockOptions(mode(FileLockManager.LockMode.None)) // Don't need to lock anything until we use the caches
+                .withCleanup(createCleanupAction(cacheMetaData))
                 .open();
     }
 
+    private CleanupAction createCleanupAction(ArtifactCacheMetadata cacheMetaData) {
+        long maxAgeInDays = DEFAULT_MAX_AGE_IN_DAYS_FOR_EXTERNAL_CACHE_ENTRIES;
+        return CompositeCleanupAction.builder()
+                .add(cacheMetaData.getExternalResourcesStoreDirectory(),
+                    new FixedAgeOldestCacheCleanup(new SingleDepthFilesFinder(ExternalResourceFileStore.FILE_TREE_DEPTH_TO_TRACK_AND_CLEANUP), maxAgeInDays))
+                .add(cacheMetaData.getFileStoreDirectory(),
+                    new FixedAgeOldestCacheCleanup(new SingleDepthFilesFinder(ArtifactIdentifierFileStore.FILE_TREE_DEPTH_TO_TRACK_AND_CLEANUP), maxAgeInDays))
+                .add(cacheMetaData.getMetaDataStoreDirectory(),
+                    new FixedAgeOldestCacheCleanup(new SingleDepthFilesFinder(ModuleMetadataStore.FILE_TREE_DEPTH_TO_TRACK_AND_CLEANUP), maxAgeInDays))
+                .build();
+    }
+
```

Comment:
It would be much nicer if `DefaultCacheLockingManager` would not have to know about the file stores that use the cache directory in this way. However, it is build-session-scoped, and the file stores are build-scoped so I didn't find a better solution.

@bigdaz Do you have an idea?

Reply:
FYI Marc and I discussed this on a call and I didn't want to add a big refactoring of the dependency management caches on top of this already big change. I've you're alright with this, I'd like to leave it up to the dependency management team to decide if, when and how this should be restructured.